Add nanostack support for KW41Z

mmahadevan108 opened the pull request:
Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

### Description
Add Thread support for FRDM KW41Z board. This was tested using the [mesh-minimal](https://github.com/ARMmbed/mbed-os-example-mesh-minimal) application. 
<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/guidelines.html#workflow (Pull request template)
-->


### Pull request type

<!-- 
    Required
    Please tick only one of the following types. Do not tick more or change the layout.

    [ ] Fix
    [ ] Refactor
    [ ] New target
    [X] Feature
    [ ] Breaking change
-->

[ ] Fix  
[ ] Refactor  
[ ] New target  
[X] Feature  
[ ] Breaking change

mmahadevan108 committed:
Add nanostack support for KW41Z

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

0xc0170 commented on features/nanostack/FEATURE_NANOSTACK/targets/TARGET_KW41Z/NanostackRfPhyKw41z.cpp:
```diff
@@ -0,0 +1,1089 @@
+#define MAC_PACKET_SIZE 127 //MAX MAC payload is 127 bytes
+static uint8_t PHYPAYLOAD[MAC_PACKET_SIZE];
+
+//TODO: verify these values
```
What does this mean "verify these values"

0xc0170 commented on features/nanostack/FEATURE_NANOSTACK/targets/TARGET_KW41Z/NanostackRfPhyKw41z.cpp:
```diff
@@ -0,0 +1,1089 @@
+    }
+}
+
+static void rf_mac_set_pending(uint8_t status) {
```
`{` new line

0xc0170 commented on features/nanostack/FEATURE_NANOSTACK/targets/TARGET_KW41Z/NanostackRfPhyKw41z.h:
```diff
@@ -0,0 +1,33 @@
+#ifndef NANOSTACK_PHY_KW41Z_H_
+#define NANOSTACK_PHY_KW41Z_H_
+
+#include "mbed.h"
```
Include what is required here, not `mbed.h`

cmonr commented on targets/targets.json:
```diff
@@ -573,7 +573,7 @@
         "core": "Cortex-M0+",
         "supported_toolchains": ["ARM", "GCC_ARM", "IAR"],
-        "extra_labels": ["Freescale", "MCUXpresso_MCUS", "KSDK2_MCUS", "FRDM"],
+        "extra_labels": ["Freescale", "MCUXpresso_MCUS", "KSDK2_MCUS", "FRDM", "FRAMEWORK_5_3_3"],
```
I take it this is the first device to use this label?
Is this label meant to be some sort of placeholder? I think we coud probably name it something a bit more descriptive.

SeppoTakalo replied:
I also agree. If you cannot find anything related to this by searching Google with `FRAMEWORK_5_3_3` then the label is not descriptive enough.

mmahadevan108 replied:
This is related to MCUXpresso SDK radio driver version.

SeppoTakalo replied:
Do we need to have multiple ones in the tree?

Or can it be combined with other flags?

Or can it be made more verbatim like "MCUXPRESSO_FRAMEWORK_5_3_3"

JuhPuur commented:
@mmahadevan108 is listed as being able to read the mac tester repository. Apparently these tests have already been used before? They are the tests in question. I wasn't aware that they had been shared.

Reverting to the commit before the transition to the new test tool removes the external dependency and the tests can be used.

cmonr commented:
@artokin Your comment about the fix coming into 5.9.0-rc1, how soon will it be coming in? The sooner it gets in, the sooner this PR can move forward.

cmonr commented:
@mmahadevan108 @SeppoTakalo @artokin Have there been any updates on bringing the patch into mbed-os? Who is taking ownership for bringing it in?

maclobdell commented:
@mmahadevan108 - can you checkout the https://github.com/ARMmbed/nanostack-mac-tester, go back a few commits in that repo until you get to Jan 23 commit (which is the one before the transition to IcedTea), then update the mbed-os repo to your local fork.  Run the tests and let us know the result. 

 @JuhPuur - can you confirm this is what you are asking Mahesh to do?

@JuhPuur  - can you push a tag or create a branch at the correct point to make this easier?  perhaps we need a feature-legacy and a feature-icedtea branch in that repo.

JuhPuur commented:
@maclobdell - confirmed. Also, good idea.

adbridge commented:
@mmahadevan108 Any updates on this? The code freeze for 5.9 is fast approaching...

cmonr commented:
Closing this PR due to lack of activity. Feel free to reopen once updates available.

mmahadevan108 commented:
The suggestion was to run the tests by going back to a Jan 23 commit. I see  the below exception when 
running against the Jan 23rd commit. I am not clear what is the way forward with this commit. 

@maclobdell Kindly suggest how we should proceed.

_________________________________________________________
Exception in thread Thread-13:
Traceback (most recent call last):
  File "C:\Python27\lib\threading.py", line 801, in __bootstrap_inner
    self.run()
  File "C:\Python27\lib\threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "C:\mbed_nanostack\nanostack-rf-driver-tester.git\test\testcases\ED_scan.py", line 66, in spam_channel
    self.command("Third", "data --dst_addr 01:02:03:00:00:00:00:01 --msdu {} --msdu_length {} --wait_for_confirm false".
format(self.payload, len(self.payload)))
  File "C:\mbed_nanostack\nanostack-rf-driver-tester.git\test\GenericTestcase.py", line 185, in execute_command
    ret = self._execute_command(device_name, command, wait = True, async=async, timeout = timeout)
  File "C:\mbed_nanostack\nanostack-rf-driver-tester.git\test\GenericTestcase.py", line 201, in _execute_command
    raise TestFail("Command didn\'t receive response \"{}\"".format(request.cmd))
TestFail: 'Command didn\'t receive response "data --dst_addr 01:02:03:00:00:00:00:01 --msdu 0123456789012345678901234567
8901234567890123456789 --msdu_length 50 --wait_for_confirm false"'
__________________________________________________

JuhPuur commented:
@mmahadevan108 That exception happens when the command does not return a return code within 50 seconds. It's curious that that's the command it fails on. Is the radio ok on that board?

mmahadevan108 commented:
@JuhPuur . I am able to run the mbed [mesh-minimal example](https://github.com/ARMmbed/mbed-os-example-mesh-minimal) on these boards.

JuhPuur commented:
@mmahadevan108 confirmed that there is some issue with the old testing framework. Had the test fail on the same error. Very annoying.

mmahadevan108 commented:
@maclobdell Kindly suggest how we should proceed.

maclobdell commented:
I'm escalating the test framework issue.

mmahadevan108 commented:
Any update on this. We would like to get this support added to mbed-os

TuomoHautamaki commented:
There are actions ongoing to get the test framework to mbed OS. Once it's there, we can start adding these tests to mbed OS as well. Summer vacation period is getting close/starting so expecting to land to mbed OS 5.10 at earliest.

FYI: @OPpuolitaival

TuomoHautamaki commented:
Hi @mmahadevan108 

Our sincere apologies as this has been on hold for too long time. We are now finally releasing the tests and test enablers [(7996),](https://github.com/ARMmbed/mbed-os/pull/7996) which would help to debug the remaining issues.
We'd really much like to get this KW41Z support added so I propose we re-open this PR and finalise. Would you be able to continue working on this? We'd be happy to support.

Regards,
Tuomo

cmonr commented:
Feel free to ping @ARMmbed/mbed-os-maintainers if this PR ends up needing to be reopened, and y'all are unable to.

mmahadevan108 commented:
I am trying to run the tests on KW41Z and running into the below issue:
`
PS C:\mymbed\mbed-os> mbed test --compile --icetea -t ARM -m KW41Z -DICETEA_MAC_TESTER_ENABLED --test-config tools\test_
configs\NanostackMACTester.json -n address_read_and_write,send_data,send_data_indirect,send_large_payloads,create_and_jo
in_PAN,ED_scan
[mbed] Working path "C:\mymbed\mbed-os" (program)
[Warning] Test with name 'address_read_and_write' was not found in the available tests
[Warning] Test with name 'send_data' was not found in the available tests
[Warning] Test with name 'send_data_indirect' was not found in the available tests
[Warning] Test with name 'send_large_payloads' was not found in the available tests
[Warning] Test with name 'create_and_join_pan' was not found in the available tests
[Warning] Test with name 'ed_scan' was not found in the available tests
`

artokin commented:
@mmahadevan108 , would you please add target `KW41Z` to the `allowed_platforms` section in the test cases available in `mbed-os/TEST_APPS/testcases/nanostack_mac_tester`? Then the application will compile in ubuntu and PC.

mmahadevan108 commented:
I have added KW41Z to the list. I had to use the below compile command:
`C:\mymbed\mbed-os> mbed test --compile --icetea -t ARM -m KW41Z -DICETEA_MAC_TESTER_ENABLED --test-config tools\test_ configs\NanostackMACTester.json -n TEST_APPS-device-nanostack_mac_tester`

Please confirm this is the right command as it is different from what is documented here:
https://github.com/ARMmbed/mbed-os/tree/master/TEST_APPS/device/nanostack_mac_tester

I also had to delete the below line from the test config file:
https://github.com/ARMmbed/mbed-os/blob/master/tools/test_configs/NanostackMACTester.json#L16

Is this fine?

artokin commented:
KW41Z RF driver should have similar `provide-default` configuration interface as described in https://github.com/ARMmbed/atmel-rf-driver/blob/master/mbed_lib.json. Then application is able to select right RF driver by using the correct `xxxx-rf.provide-default": true`. There is no need to delete the line if configuration for KW41Z is otherwise made correctly.

I think the compilation command in the documentation is correct BUT we may have an error in test case lookup.  We will investigate this issue. If the test cases can't be found by the application the application you can still use the application manually. Please see examples from the test cases.

artokin commented:
@mmahadevan108 please note Mbed OS folder structure has slightly hanged. 

- Mbed OS 802.15.4 RF drivers are now inside of Mbed OS in `mbed-os/components/802.15.4_RF` 
- `FEATURE_NANOSTACK` has been removed from the folder structure.

mmahadevan108 committed:
Update TEST_APPS to add KW41Z for nanostack tests

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
> @mmahadevan108 please note Mbed OS folder structure has slightly hanged.
> 
> * Mbed OS 802.15.4 RF drivers are now inside of Mbed OS in `mbed-os/components/802.15.4_RF`
> * `FEATURE_NANOSTACK` has been removed from the folder structure.

Thank you. I have updated this PR per the new folder structure.

artokin commented:
@mmahadevan108 , there is no problem with test case lookup, the warning message was misleading. 

After changing the default driver from `NanostackMACTester.json#L16` I was able to run `address_read_and_write` test case (as I was having only one KW41Z + MCR20A) by using the following command:

` mbed test --clean --run --compile --icetea -t ARM -m KW41Z -DICETEA_MAC_TESTER_ENABLED --test-config tools/test_configs/NanostackMACTester.json -n address_read_and_write
`

I believe the test framework is OK and all the tests can be automatically executed.

mmahadevan108 commented:
Thank you. I am able to run the tests. I am seeing the below pass
address_read_and_write,send_data,send_large_payloads,create_and_join_PAN

However the below fails: 
send_data_indirect

Any suggestions on what I might be missing that is causing this failure.

mmahadevan108 commented:
I have pushed an update that passes the send_data_indirect test. Can you please review

mmahadevan108 commented:
I am seeing the ED scan test crash with the below error:
_____________________________________
Exception in thread Thread-103:
Traceback (most recent call last):
  File "c:\python27\lib\threading.py", line 801, in __bootstrap_inner
    self.run()
  File "c:\python27\lib\threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "C:\mymbed\mbed-os\TEST_APPS\testcases\nanostack_mac_tester\ED_scan.py", line 64, in spam_channel
    self.lock_th.release()
error: release unlocked lock
___________________________________________

artokin commented:
There is error in the `ED_scan` test case exception handling.  Already started Thread is not stopped when rerunning the test.  We will fix that issue soon.

However, it looks like the ED_scan test is failing because the exception is caught. Would you have an update for that issue as well?

mmahadevan108 commented:
I will rerun the test once this fix is merged in.

kjbracey commented on features/nanostack/targets/TARGET_NXP/TARGET_KW41Z/NanostackRfPhyKw41z.cpp:
```diff
@@ -0,0 +1,1093 @@
+    platform_exit_critical();
+}
+
+NanostackRfPhy &NanostackRfPhy::get_default_instance()
```
I think this should really be `MBED_WEAK` - for the PHYs we have the model that there's no default implementation in the OS, the target can supply a weak implementation, and the app or a plug-in module can provide normal.

This allows the theoretical possibility of plugging in a phy module into a board that has a built-in phy.

Can be fixed later though.

kjbracey reviewed (approved):
Looks good to me, looking at the radio driver at least. Not familiar with the test config, and not studied the platform code.

mmahadevan108 commented:
I reran the tests with the fix in https://github.com/ARMmbed/nanostack-mac-tester/pull/27, all tests passed,

+------------------------+---------+-------------+-------------+-----------+----------+
| Testcase               | Verdict | Fail Reason | Skip Reason | platforms | duration |
+------------------------+---------+-------------+-------------+-----------+----------+
| ED_scan                |   pass  |             |             |   KW41Z   | 151.173  |
| address_read_and_write |   pass  |             |             |   KW41Z   |  2.786   |
| create_and_join_PAN    |   pass  |             |             |   KW41Z   |  11.276  |
| send_data              |   pass  |             |             |   KW41Z   |   4.75   |
| send_data_indirect     |   pass  |             |             |   KW41Z   |  8.179   |
| send_large_payloads    |   pass  |             |             |   KW41Z   |  18.767  |
+------------------------+---------+-------------+-------------+-----------+----------+
+---------------+----------------+

kseverinkangas-zg commented on features/nanostack/targets/TARGET_NXP/TARGET_KW41Z/NanostackRfPhyKw41z.cpp:
```diff
@@ -0,0 +1,1093 @@
+    xcvrStatus_t xcvrStatus;
+    uint32_t phyReg;
+
+    xcvrStatus = XCVR_Init(ZIGBEE_MODE, DR_500KBPS);
```
Clarification needed here, what kind of configuration this actually creates. Thread uses 256kbps rate.

mmahadevan108 replied:
This helps setup the hardware, below are the data rate options available:
https://github.com/NXPmicro/mbed/blob/Add_KW41Z_Thread/targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/middleware/wireless/TARGET_FRAMEWORK_5_3_3/XCVR/MKW41Z4/fsl_xcvr.h#L445

kseverinkangas-zg replied:
For Thread the data rate should be 256kbps. I suppose ZIGBEE_MODE refers to 802.15.4 spec?
Are you planning to support other standards with this driver?

mmahadevan108 replied:
Yes ZIGBEE_MODE is 802.15.4. There are no plans to support other standards with this driver.

kseverinkangas-zg commented on features/nanostack/targets/TARGET_NXP/TARGET_KW41Z/NanostackRfPhyKw41z.cpp:
```diff
@@ -0,0 +1,1093 @@
+        if (framePending) {
+            device_driver.phy_tx_done_cb(rf_radio_driver_id, rf_mac_handle, PHY_LINK_TX_DONE_PENDING, 1, 1);
+        } else {
+            // arm_net_phy_tx_done(rf_radio_driver_id, mac_tx_handle, PHY_LINK_TX_SUCCESS, 1, 1);
```
can be removed

mmahadevan108 replied:
I have updated the PR to remove this line.

0xc0170 commented on features/nanostack/targets/TARGET_NXP/TARGET_KW41Z/NanostackRfPhyKw41z.h:
```diff
@@ -0,0 +1,33 @@
+#ifndef NANOSTACK_PHY_KW41Z_H_
+#define NANOSTACK_PHY_KW41Z_H_
+
+#include "mbed.h"
```
please include what is needed, not `mbed.h`

mmahadevan108 replied:
I got this from other supported targets, 
https://github.com/ARMmbed/mbed-os/blob/master/features/nanostack/targets/TARGET_NCS36510/NanostackRfPhyNcs36510.h#L20
https://github.com/ARMmbed/mbed-os/blob/master/features/nanostack/targets/TARGET_Silicon_Labs/TARGET_SL_RAIL/NanostackRfPhyEfr32.h#L20

mmahadevan108 commented:
Please let me know if anything more is needed.

mmahadevan108 commented:
Can the change related to include "mbed.h" be addressed as part of a separate PR as it is not limited to this driver.

kseverinkangas-zg commented:
Only one thing: DR_500KBPS  to DR_256KBPS ?

0xc0170 commented:
> Can the change related to include "mbed.h" be addressed as part of a separate PR as it is not limited to this driver.

It should be quick to replace ? We are cleaning this up, so would like to not have it in. 

Looking at the header file, it should not need it, most likely some implementation file might miss some declarations.

mmahadevan108 commented:
> Only one thing: DR_500KBPS to DR_256KBPS ?

The data rate argument only matters when GFSK/MSK protocol is selected.

kseverinkangas-zg commented:
Ok, maybe a comment could be added -> only matters when GFSK/MSK protocol used.

mmahadevan108 commented:
> Ok, maybe a comment could be added -> only matters when GFSK/MSK protocol used.

Comment added

0xc0170 reviewed (approved):
Approving but will wait for nanostack kw41 fixed asap

0xc0170 commented:
> Approving but will wait for nanostack kw41 fixed asap

Note, we are going add check for any inclusion of mbed.h where it should not be used. Therefore this should be addressed soon, otherwise might be reverted if not fixed.

0xc0170 commented:
We will start CI but noticed first Travis failures - we include workaround in travis config . Can you rebase to get it in, travis should be green

mmahadevan108 commented:
> We will start CI but noticed first Travis failures - we include workaround in travis config . Can you rebase to get it in, travis should be green

Rebased and also removed include of mbed.h

0xc0170 commented:
We fixed Travis failures on master hours ago, and this change should be propagated here via rebase. We will start CI right after travis fix.

I can rebase this later today if not yet done. Sorry for the trouble

mmahadevan108 commented:
@0xc0170 I have rebased this PR

0xc0170 commented:
CI started

mmahadevan108 commented:
> # [Test run: FAILED](https://mbed-os.mbedcloudtesting.com/job/mbed-os-ci/job/PR-6622/1/)
> Summary: 1 of 11 test jobs failed
> Build number : 1
> [Build artifacts](http://mbed-os-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/artifacts/6622/)
> 
> Failed test jobs:
> 
> * jenkins-ci/mbed-os-ci_greentea-test

I am not clear how to address the network test failures seen.

SeppoTakalo commented:
@mmahadevan108 By default the Greentea tries to run network tests for all boards that provide default network interface, as this does by default.

However, our test environment does not (yet) have properly working Mesh network available. Therefore you need to tell Greentea run that by default, use NO_NETWORK test configuration.

See example here for `KW24D` platform: https://github.com/ARMmbed/mbed-os/blob/master/tools/test_configs/target_configs.json#L54..L57

This NO_NETWORK disables the default network interface for Greentea builds, until `mbed_app.json` is provided. This leads to build of DNS and network interface tests being skipped.

Once you copy&paste the `KW24D` section for your board, the test run should pass.

mmahadevan108 committed:
Supply test config for KW41Z

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
Thank you. I have updated this PR with the change suggested.

mmahadevan108 commented:
@0xc0170 Anything needed for this PR from my end?

0xc0170 commented:
CI started
